cmd/snap-confine/snap-confine.c

```diff
@@ -414,18 +420,49 @@ int main(int argc, char **argv)
 	// For classic and confined snaps
 	sc_selinux_set_snap_execcon();
 #endif
-	if (sc_apply_seccomp_profile_for_security_tag(invocation.security_tag)) {
-		/* If the process is not explicitly unconfined then load the global
-		 * profile as well. */
-		sc_apply_global_seccomp_profile();
-	}
 	if (snap_context != NULL) {
 		setenv("SNAP_COOKIE", snap_context, 1);
 		// for compatibility, if facing older snapd.
 		setenv("SNAP_CONTEXT", snap_context, 1);
 	}
+
+	// Normally setuid/setgid not only permanently drops the UID/GID, but
+	// also clears the capabilities bounding sets (see "Effect of user ID
+	// changes on capabilities" in 'man capabilities'). To load a seccomp
+	// profile, we need either CAP_SYS_ADMIN or PR_SET_NO_NEW_PRIVS. Since
+	// NNP causes issues with AppArmor and exec transitions in certain
+	// snapd interfaces, keep CAP_SYS_ADMIN temporarily when we are
+	// permanently dropping privileges.
+	if (getresuid(&real_uid, &effective_uid, &saved_uid) != 0) {
+		die("getresuid failed");
+	}
+	debug("ruid: %d, euid: %d, suid: %d",
+	      real_uid, effective_uid, saved_uid);
+	struct __user_cap_header_struct hdr = { _LINUX_CAPABILITY_VERSION_3, 0 };
+	struct __user_cap_data_struct data[2] = { {0} };
+
+	// At this point in time, if we are going to permanently drop our
+	// effective_uid will not be '0' but our saved_uid will be '0'. Detect
+	// and save when we are in the this state so know when to setup the
+	// capabilities bounding set, regain CAP_SYS_ADMIN and later drop it.
+	bool keep_sys_admin = effective_uid != 0 && saved_uid == 0;
+	if (keep_sys_admin) {
+		debug("setting capabilities bounding set");
+		// clear all 32 bit caps but SYS_ADMIN, with none inheritable
+		data[0].effective = CAP_TO_MASK(CAP_SYS_ADMIN);
+		data[0].permitted = data[0].effective;
+		data[0].inheritable = 0;
+		// clear all 64 bit caps
+		data[1].effective = 0;
+		data[1].permitted = 0;
+		data[1].inheritable = 0;
+		if (capset(&hdr, data) != 0) {
```

Comment:
```suggestion
		if (capset(&hdr, data) < 0) {
```

Reply:
This falls into the category of https://github.com/snapcore/snapd/pull/7026#discussion_r297842904 where I find the '!= 0' clearer here (since 0 is the only measure of success).